honeybee_grasshopper_core/src/HB Apertures by Ratio.py

```diff
@@ -0,0 +1,209 @@
+        ep_constr_: Optional text for an energy construction to be used for all
+            generated apertures. This text will be used to look up a construction
+            in the window construction library. This can also be a custom
+            WindowConstruction object. If an array of text or construction objects
+            are input here, different constructions will be assigned based on
+            cardinal direction, starting with north and moving clockwise.
+        rad_mat_: Optional text for a radiance material to be used for all
+            generated apertures. This text will be used to look up a material
+            in the material library. This can also be a custom material object.
+            If an array of text or material objects are input here, different
+            materials will be assigned based on cardinal direction, starting
+            with north and moving clockwise.
+        _run: Set to True to run the component.
+    
+    Returns:
+        report: Reports, errors, warnings, etc.
+        hb_obj: The input Honeybee Face or Room with Apertures generated from
+            the input parameters.
+"""
+
+ghenv.Component.Name = "HB Apertures by Ratio"
+ghenv.Component.NickName = 'AperturesByRatio'
+ghenv.Component.Message = '0.1.0\nOCT_28_2019'
+ghenv.Component.Category = "HoneybeeCore"
+ghenv.Component.SubCategory = '0 :: Create'
+ghenv.Component.AdditionalHelpFromDocStrings = "4"
+
+try:  # import the core honeybee dependencies
+    from honeybee.boundarycondition import Outdoors
+    from honeybee.facetype import Wall
+    from honeybee.room import Room
+    from honeybee.face import Face
+    from ladybug_rhino.config import tolerance, conversion_factor
+    from ladybug_rhino.grasshopper import all_required_inputs
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee:\n\t{}'.format(e))
+
+try:  # import the honeybee-energy dependencies
+    from honeybee_energy.lib.constructions import window_construction_by_name
+    from honeybee_energy.construction.window import WindowConstruction
+except ImportError:
+    pass  # honeybee-energy is not installed and ep_constr_ will not be avaialble
+
+
+def can_host_apeture(face):
+    """Test if a face is intended to host apertures (according to this component)."""
+    return isinstance(face.boundary_condition, Outdoors) and \
+        isinstance(face.type, Wall)
+
+
+def assign_apertures(face, sub, rat, hgt, sil, hor, vert, op, ep, rad):
+    """Assign apertures to a Face based on a set of inputs."""
+    if sub:
+        face.apertures_by_ratio_rectangle(rat, hgt, sil, hor, vert, tolerance)
+    else:
+        face.apertures_by_ratio(rat, tolerance)
+    
+    # try to assign the operable property
+    if op:
+        for ap in face.apertures:
+            ap.is_operable = op
+    
+    # try to assign the energyplus construction
+    if ep is not None:
+        for ap in face.apertures:
+            ap.properties.energy.construction = ep
```

Comment:
I disagree with this since it makes it extremely difficult to assign constructions for only the new apertures, shades, etc. generated from a given component. What you are proposing essentially forces a user to re-assign all the constructions of all of the apertures of a Room (since there's not necessarily a way for a component that comes after this one to know which ones are new;y generated vs. existing before the component). 
All of this is to say that it's going to take a lot to convince me that this isn't a case where practicality beats purity.  An extra 3 lines of code in order to support important functionality is well worth it in my mind.

Reply:
I agree with your comment here. That was a generic comment about the approach which we are discussing in a different thread.